### ISSUE TITLE
Support for Ranking scenario

### ISSUE THREAD

**Issue by torronen** [dotnet/machinelearning-modelbuilder#2108]
**Is your feature request related to a problem? Please describe.**
Ranking is one of the features of ML.NET. Model Builder is missing this scenario.
I found only mention of it here https://github.com/dotnet/machinelearning-modelbuilder/issues/702 but perhaps it ha been delayed since then.

**Describe the solution you'd like**
I'd like support for Ranking scenario.

**Describe alternatives you've considered**
Currently, I am trying to run ranking through AutoML C# API, but having trouble (always get NDCG = 1.0 in every iteration)

**Comment by beccamc** [comment]
Thanks for adding this issue! We haven't had a lot of requests for this scenario. We'd love to learn more about your use case... can you give us some more details on what you want to accomplish? 

@LittleLittleCloud Have you tried this?

**Comment by torronen** [comment]
My current use case is to enhance a chatbot classifier. If NN seems uncertain then we could show list of most relevant topics. Current data is numbers only (embeddings + some other data).

> Sorry, I did not understand. Would you like to know more about one of these topics:
> [1. Highest ranking]
> [2. 2nd highest ranking]

It is probably correct there are less use cases for ranking in general. I am not sure, but maybe, most classification tasks could be reframed as ranking tasks (probability of belonging to a class vs. what is the most relevant class). I have no idea yet if this would be better or not. Do you know if it could be possible ranking gives better results, or should it basically give the same results? Maybe recommendation tasks could be ranking tasks, too e.g. "what is the highest ranking product for me" vs. "which products to recommend". 

I do not have code example yet. My best guess is GroupId is not getting set correctly and it might be only 1 item per group. These are the most important lines from my current attempt (modified from AutoML classification)
```
               columnInference.ColumnInformation.GroupIdColumnName = GroupColumn;
                var experiment = mlContext.Auto().CreateRankingExperiment(experimentSettings);
                experimentResult = experiment.Execute(TrainDataView, ValidationDataView, columnInformation, null, progressHandler);
```

**Comment by LittleLittleCloud** [comment]
You can create your own automl experiment using sweepable api now! It allows you to optimize parameter on customed pipeline and search space and I can help you on this task, would you mind sharing subset of your dataset to begin with?

**Comment by torronen** [comment]
@LittleLittleCloud This is one validation set I have been trying https://u.pcloud.link/publink/show?code=XZ1PppVZfy8O5LuUeQFf6HKUpyeyBXxNkND7  

Columns Rank and Group are the last ones. Other columns are scores and embeddings from models trying to match a query to topics. Distribution by rank:
4 - 749
3 - 3000
2 - 7286
1 - 250

I am still having the problem of getting 1.0 NDCG for every position I tried (1,5,20,50 etc)
```
columnInference.ColumnInformation.GroupIdColumnName = GroupColumn;
var experiment = mlContext.Auto().CreateRankingExperiment(experimentSettings);
experimentResult = experiment.Execute(TrainDataView, ValidationDataView, columnInformation, null, progressHandler);
```
and modification of https://github.com/dotnet/machinelearning-samples/tree/main/samples/csharp/getting-started/Ranking_Web


I think I will next try with LightGBM executable to confirm it works.

**Comment by LittleLittleCloud** [comment]
One tip when running ranking task: you probably want to sort according to GroupId before feeding dataset into pipeline because ML.Net doesn't respect group id sometimes

check the following for detailed information
https://github.com/dotnet/machinelearning/issues/5022

**Comment by torronen** [comment]
Wow, thanks, that is cool! I suppose my problem has been not sorting by GroupId. Actually, our datasets are the opposite. There was some reason why datasets with similar items one after another caused problems (maybe something with Keras models?). The current datasets were created so that similar rows are not one after each other, so instead it is 1 item from each group, then 2nd from each group etc.

Is there a way "inject" potentially best parameter sets to run first? It does not matter for this use case but for big datasets I may have only 20-50 iterations of time budget available and I'd like to try with the best ones from similar runs. We have files like this: 
[params-FastTreeRanking-0000010-c6a4e8b6d3374611e108e87b8f65d2b0577dd9a5f677c510723c5d702dd6a1c0.json.txt](https://github.com/dotnet/machinelearning-modelbuilder/files/8619966/params-FastTreeRanking-0000010-c6a4e8b6d3374611e108e87b8f65d2b0577dd9a5f677c510723c5d702dd6a1c0.json.txt)
along the .zip model. List of metrics are in the database. 

So for example, ideally, I'd start the experiment with the above values. Or, is it better to first run separately with these values, and then perhaps use the results to edit the search space?

**Comment by torronen** [comment]
Another question I have: how to change the metric to optimize? How to create a custom metric? Custom metric is sometimes needed because not all entries are equal. For example, it is better to detect few severe anomalies than all light anomalies, even if standard metrics would be lower because of it.

I'll have a look at the API later to understand better, questions just FYI.

**Comment by LittleLittleCloud** [comment]
> Is there a way "inject" potentially best parameter sets to run first?
Yeah you can, it would need to be in two `AutoMLExperiment` though, where the first experiment would go through a set of pre-defined parameters and the second one go through original search space. 

To achieve this you would need to provide a customize search space. The gist of this is to use a `Choice` option which maps to pre-defined parameter sets.
```csharp
// pre-defined parameters
class MyOption
{
    // suppose you have five parameters to try first
    [Choice(0,1,2,3,4)]
    public int ParameterIndex {get; set;}
}
```
Then in `CreateSweepableEstimator`, simply create ml.net trainer according to `MyOption.ParameterIndex`

> how to change the metric to optimize? How to create a custom metric?
You can calculate your own metric and pass it to `AutoMLExperiment` through `TrialResult` in `ITrialRun.Run` method. (in L31) 
![image](https://user-images.githubusercontent.com/16876986/166744331-5d14196b-8f19-4929-aabe-8077e3face3d.png)
